chain/client/src/client.rs

```diff
@@ -919,37 +945,48 @@ impl Client {
     /// * `is_ours`  - whether the approval was just produced by us (in which case skip validation,
     ///                only check whether we are the next block producer and store in Doomslug)
     pub fn collect_block_approval(&mut self, approval: &Approval, is_ours: bool) {
-        let Approval { parent_hash, account_id, parent_height, target_height, signature } =
-            approval;
+        let Approval { inner, account_id, target_height, signature } = approval;
 
         let process_error = |e: near_chain::Error,
                              approval: &Approval,
                              pending_approvals: &mut SizedCache<_, _>| {
             if let ErrorKind::DBNotFoundErr(_) = e.kind() {
                 let mut entry = pending_approvals
-                    .cache_remove(&approval.parent_hash)
+                    .cache_remove(&approval.inner)
                     .unwrap_or_else(|| HashMap::new());
                 entry.insert(approval.account_id.clone(), approval.clone());
-                pending_approvals.cache_set(approval.parent_hash, entry);
+                pending_approvals.cache_set(approval.inner.clone(), entry);
             }
         };
 
-        let next_epoch_id =
-            match self.runtime_adapter.get_epoch_id_from_prev_block(&approval.parent_hash) {
-                Err(e) => {
-                    process_error(e, approval, &mut self.pending_approvals);
-                    return;
+        let parent_hash = match inner {
+            ApprovalInner::Endorsement(parent_hash) => parent_hash.clone(),
+            ApprovalInner::Skip(parent_height) => {
```

Comment:
what prevents someone from sending a lot of such skips and saturate the cache?

Reply:
This cache is generally not sybil resistent (they can also send endorsements with fake prev hash).
Hopefully the endorsement delay is sufficient generally for the block to propagate, and this cache is not playing a crucial role.